# Introduction

During my first internship I got the chance to develop a package which determines the ploidy levels (major, minor and copynumber) per autosome arm for metastatic tumour samples, and a package to analyze these ploidy levels. It was a great opportunity to get more insight in R and the biology behind structural autosome changes in metastatic cancer. 

# ploidyEstimator

This script determines the major, minor and total copynumber (ploidy levels) per autosome arm for metastatic cancer samples. Returns three .tsv files with per sampleID the ploidy levels of each autosome arm and the overall ploidy level of the sample (genome_level).

USE:  ploidyEstimator(path = "C://Documents/...", files = filenames)

### Input

- Path to .cnv files

A .cnv file with the following information is used as input for the ploidy estimator:

```
#chromosome	start	end	copyNumber	bafCount	observedBAF	actualBAF	segmentStartSupport	segmentEndSupport	method
```

### Output

The three .tsv output files all have the same layout as displayed below:

```
"chromArms"	  "sampleID#1"	  "sampleID#2"	  "sampleID#3"  "sampleID#4"	  "sampleID#5"	
"1p"	            2                 1               1               3                 2
"1q"	            1                 1               1               2                 2
"2p"	            1                 2               1               2                 3
"2q"	            2                 2               2               4                 1
....
.etc
....
"22q"	            1                 1               2               2                 2
"genome_level"      1                 1               1               2                 2
```

## Dependencies

* "data.table" package

## Versioning

* Version 1.0 - 09-10-2018

* Version 1.1 - 08-11-2018
  * Added getChromArm function
  
* Version 2.0 - 03-12-2018
  * Divided script into different files
  
* Version 3.0 - 10-12-2018
  * Completed package version of ploidyEstimator
  
# ploidyAnalyzer

The ploidy analyser is a package which is made to analyze the ploidy levels generated by the ploidy estimator. It generates multiple diagnostical plots as output. 

### Input

The input for this script is the output from the aneuploidyEstimator and some annotation files:
* mut_matrix_and_mut_load.csv
* mut_matrix_and_mut_load_clonal_subclonal.csv
* mut_matrix_and_sign_contribution.csv
* HMF_Annotations.txt
* msi_patients.txt

The annotation files contain information about, primary tumor location, microsatellite instability,  mutational load (clonal & subclonal) and signature contributions.

### Output

The output contains out of .pdf and .png files with diagnostic plots (heatmaps, barplots, boxplots, violin plots etc.). 

## Dependencies

Packages:
* lattice
* stringr
* ggplot2
* heatmap.plus
* reshape2

## Versioning

* Version 1.0 - 20-11-2018
* Version 1.1 - 28-11-2018
  * Added functions for unsupervised heatmaps
* Version 1.2 - 29-11-2018
  * Added violinAneuploidy function for violin plots with gain and loss distribution
* Version 1.3 - 12-12-2018
  * Added "ghost" function extraPlotFunctions.R for plot script which aren't in use at this moment
* Version 2.0 - 17-12-2018
  * Completed package version of ploidyAnalyzer
 

## Authors

* **M. Opperman** 

## Acknowledgement

I would like to thank the whole Cuppen group, especially Edwin Cuppen, for a great first internship experience and teaching me a lot about biology and bioinformatics.
A special thank you to Arne van Hoeck who was my direct supervisor, and Luan Nguyen for helping me get better at R.  


